spec/filtering_spec.rb

```diff
@@ -174,134 +186,120 @@
     end
   end
 
-  context 'when the filter is guarded' do
-    let(:can_filter) { true }
-    let(:ctx) { double(can_filter_first_name?: can_filter).as_null_object }
-
+  context 'when filtering on an unknown attribute' do
     before do
-      params[:filter] = { first_name_guarded: 'Stephen' }
+      params[:filter] = { foo: 'bar' }
     end
 
-    context 'and the guard passes' do
-      it 'filters normally' do
-        JsonapiCompliable.with_context(ctx, {}) do
-          expect(records.map(&:id)).to eq([employee1.id])
-        end
-      end
+    it 'raises helpful error' do
+      expect {
+        records
+      }.to raise_error(JsonapiCompliable::Errors::AttributeError, 'AnonymousResourceClass: Tried to filter on on attribute :foo, but could not find an attribute with that name.')
     end
 
-    context 'and the guard does not pass' do
-      let(:can_filter) { false }
+    context 'but there is a corresponding extra attribute' do
+      before do
+        resource.extra_attribute :foo, :string
+      end
 
-      it 'raises an error' do
-        expect {
-        JsonapiCompliable.with_context(ctx, {}) do
+      context 'but it is not filterable' do
+        it 'raises helpful error' do
+          expect {
             records
-          end
-        }.to raise_error(JsonapiCompliable::Errors::BadFilter)
+          }.to raise_error(JsonapiCompliable::Errors::AttributeError, 'AnonymousResourceClass: Tried to filter on on attribute :foo, but the attribute was marked :filterable => false.')
```

Comment:
Not a blocker here, but would be extra helpful to have different error messages for when `ApplicationResource.default_filterable = false` versus `attribute :foo, filterable: false`